### ISSUE TITLE
[RFC] Better way to override Backstage's configuration from a Kubernetes deployment

### ISSUE THREAD

**Issue by martina-if** [spotify/backstage#2889]
## Need

A way to dynamically override the configuration of Backstage backend.

Today, the configuration of Backstage is loaded from a set of files with predefined names and locations in the [resolver.ts](https://github.com/spotify/backstage/blob/master/packages/config-loader/src/lib/resolver.ts#L28). From that file:

> Resolves all configuration files that should be loaded in the given environment.
>   
> For each root directory, search for the default app-config.yaml, along with suffixed
>  APP_ENV and local variants, e.g. app-config.production.yaml or app-config.development.local.yaml
>   
> The priority order of config loaded through suffixes is `env > local > none`, meaning that
>  for example app-config.development.yaml has higher priority than `app-config.local.yaml`. 

After #2845 was merged we also have the extra flexibility of defining `APP_ENV` environment variable to
decide which config file will be selected with the highest priority.

I see two issues today with this way to configure Backstage in Kubernetes:

- First, mounting an extra configuration file (e.g.: `app-config.production.yaml`) can't disable things that have been enabled in the default config `app-config.yaml`
- Second, to overcome this, the solution is to mount a file called `app-config.yaml` on top of the same file that lives inside the docker image which is not great.

I think it would be nicer to not have to place this file in a specific location and potentially overriding one of the existing config files (inside the docker image). Instead, we could mount the config file anywhere (e.g: `/etc/backstage/app-config.yaml`) and then point the Backstage application to where the file is.

## Proposal

This would work in a similar way as other unix tools. For example in the docker daemon:

```
$ dockerd --help

Usage:	dockerd [OPTIONS]

A self-sufficient runtime for containers.

Options:
      ...
      --config-file string                      Daemon configuration file (default "/etc/docker/daemon.json")

```

Or dnsmasq:

```
$ dnsmasq --help
Usage: dnsmasq [options]

Valid options are:
...
-C, --conf-file=<path>                                 Specify configuration file (defaults to /etc/dnsmasq.conf).
```

An important aspect to this is that this file should completely override the other config files. Today, the cascade of config files work as a "fallback" system, so that if a field is `null` in a higher priority file, the low priority file will fill in with its value. This means there is no way to disable configuration already existing in a docker image.

If one has the default app-config.yaml:
```
scaffolder:
  azure:
    baseUrl: https://dev.azure.com/{your-organization}
    api:
      token:
        $env: AZURE_TOKEN
```

and wants to disable azure, even if there is another file called `app-config.local.yaml` with

```
scaffolder:
  azure: null
```

The azure scaffolder won't be disabled, and the backend won't be able to start unless a valid `baseUrl` and `token` are provided.

## Alternatives

1. The location of this config file could be provided via a CLI argument or perhaps through an environment variable.

2. Another alternative is to continue using the current system and always override `app-config.yaml` from the Kubernetes `deployment`.

3. To have a very minimal `app-config.yaml` and then have a second `app-config.<APP_ENV>.yaml` file that enable the different plugins and options in an opt-in basis.

4. To allow disabling parts of the `app-config.yaml` file through `null` or similar values placed in `app-config.<APP_ENV>.yaml`.

I am sure there are more alternatives to consider, please feel free to add more in the comments.

## Risks

With 1 I see the risk of getting an overly complicated system for configuration loading if it is also combined with the existing one. Perhaps implementing 1 should mean having a simplified version of the existing system where there is only the options of `app-config.yaml`, `app-config.development.yaml` and the complete override with `--config my-app-config.yaml`.

Alternative 3 sounds also very compelling to me because most of the configuration can be just moved to the development config file and have a very minimal and clean default config. This would mean the least amount of changes while keeping the flexibility of the "fallback" system.

**Comment by Rugvip** [comment]
👍 , been meaning to create an issue around the `null` behaviour in overrides. Happy for contributions or creation of a more specific issue if we don't get around to it first.

Regarding config loading, there are a couple of things I feel are really wonky with the current setup:

1. It is not obvious **when** config is loaded, and it's easy to make the mistake of for example trying to configure frontend values in a deployment `app-config.yaml`.
1. It's easy to get surprised by additional config files being loaded into docker images and picked up unintentionally.
1. Providing additional configuration at deployment time is awkward, since there's a limited number of files to work with.

This issue covers (3), but there may be an opportunity the tackle (1) and (2) too. What do you think about the following @martina-if? building on top of your idea of adding `--config-file`:

- Remove env-based config, reading of `APP_ENV` and `NODE_ENV` are removed.
- Add a `--config` flag to relevant `backstage-cli` commands. Either required, or defaulted to a root `app-config.yaml`.
- Add a top-level `extends` field to config, which will load the extended config first. This would make it simple to emulate the existing behaviour, but provides much more flexibility in naming and location of config. An option to the could be to allow multiple `--config` flags, but tbh the DX of that isn't great imo.

The intention that this would help out a bit with (1), because having the flag provides more visibility, and you simply can't provide a `--config` flag to e.g. `backstage-cli app:build` inside an already built image. It would eliminate (2) as a problem, because you have to explicitly point to your config file. It also address (3), which ofc is what the original suggestion by @martina-if  is about.

Couple of problems with this approach though:

- Including config files in the `build-image` command for the backend is less straight forward. Currently we just include all possible `app-config` yamls, but with custom names we can't copy in the config files anymore.
- `app-config.local.yaml` is pretty convenient and without anything extra added to the above solution it wouldn't be supported. For example, for any `<name>.yaml` config file that is specified, we could load in any `<name>.local.yaml`, but that could again be a bit too much magic.

**Comment by martina-if** [comment]
Thanks for the feedback and the new ideas.

When you say passing the config to  `backstage-cli app:build` is this for the frontend? Or is this needed for the backend part too?

The fact that the frontend sample app can't load dynamic configuration is a limitation that would be nice to resolve. We can always configure it through environment variables so that's good. Ideally you could be able to dynamically load any given config file just like the backend but I don't know if this is doable.

I like the option of the `extends`. It adds a lot of flexibility, on the other hand I could imagine it can become a bit repetitive to have a file with keys both in the root and under `extends:`. Another similar idea:

- having only `--config` and `--extra-config`: the first is the only config it will use. The second is an extension to the existing configuration living in the default places. This doesn't allow you to pick and choose what overrides and what "extends" but I like that it's simple

In the end, I would like something that allows us to:
- easily have a local development configuration
- easily override any config field for prod deployments

Perhaps having a tiny app-config.yaml, an extended app-config.development.yaml and adding the `--config` option for kubernetes deployments could be enough. This would be in line to what [apollo](https://github.com/spotify/apollo/tree/1.x/apollo-core#consistent-configuration) does.

**Comment by Rugvip** [comment]
Yeah config flags would need to be passed to any command that needs config, so I think it might be better to make it a required flag to avoid confusion. That's from the CLI's point of view though. Most usages will be wrapped up in yarn scripts or the like anyway.

Regarding frontend config, the `@backstage/core` package is set up so that you can provide your own config loading logic, but I think we could look into other default ways to get config to the frontend too. Would be pretty interested in trying out a simple `/.app-config` endpoint or similar, with some JSONP-type thing but just assigning to a var. I think the one thing to avoid is loading the config through a fetch, since that requires the main app to start executing first, which would add quite a bit of delay to the page load.

Regarding dynamic runtime config for the app a limitation is that that config is actually used as part of the build process to set things like the public URL in webpack. We could also look into more ways than the `APP_CONFIG_`-way of supplying runtime config, as the end result of that is just a JSON blob anyway.

Regarding `extends`, I should clarify the extends idea. It would look something like this:

```yaml
# app-config.yaml
app:
  baseUrl: http://localhost:3000

backend:
  baseUrl: http://localhost:7000
  listen:
    port: 7000
```

```yaml
# app-config.production.yaml

extends: ./app-config.yaml # loads all config first, and merges them together using the existing logic

app:
  baseUrl: https://localhost:3000

backend:
  baseUrl: https://localhost:7000
  listen:
    port: 7000
```

The docker run command would be something like

```bash
node packages/backend --config app-config.production.yaml
```

And that would emulate the current behaviour of `APP_ENV=production`.

**Comment by andrewthauer** [comment]
I think I should share a relevant story that might inform this discussion ...

After we upgraded to `alpha.25` our deploys were failing because the `backend` couldn't talk to our `postgres db` any longer. It was trying to connect to `localhost` which smelled very much like a configuration fallback issue to the `app-config.yaml`. However, I couldn't replicate the problem locally. After trying too many things to count, I ended up realizing the new `APP_ENV` was causing the issues. There were actually 2 problems. One was our prod runtime infrastructure uses a custom `APP_ENV`, and second the `frontend` build got the wrong config since our CI sets `APP_ENV=test`. So this was essentially making our previously working `app-config.production.yaml` a noop in CI & prod.

This was quite a frustrating experience to troubleshoot. I do find the current override system useful for local development, but there's a lot of magic going on and it's very hard to predict what is happening in a deterministic way. Especially when it comes to baking the configs into the frontend build. A couple of improvements come to mind here:

- Clearly output how the app-configs are being resolved, merged, etc. at both build time and when the backend starts up
- I think the idea having an explicit CLI option to force the configs(s) to be used would be good. It would at least make the resolution more deterministic.

**Comment by Rugvip** [comment]
@andrewthauer `APP_ENV` already being used for other things is ofc a thing I should've considered when adding that. Quick google gives quite a few results. Still much fewer results than `NODE_ENV` at least 😅 . Sorry about the frustration though, and hoping we can find some good solutions to avoid more of it

+1 for logging the resolved config at build time, gonna get on that straight away :grin:

Regarding adding config flags to the CLI I'm just a bit torn on two points:
1. Do we allow multiple flags, or use `extends:` in the config. Multiple flags can be a bit tedious and it is less obvious in which order the config is being merged. Logging the resolved config files is gonna help there, and we could ofc also support both. Extends is a bit more magic since we have a new special top-level keyword, which makes me hesitant to add it, and it's also tedious when working with multiple dimensions of config.
2. How do we do local overrides with flags? Feeling that the local overrides is a required feature at this point, but it is less obvious how to work it into a flag-based solution. Perhaps no flag means loading in `app-config.yaml` and `app-config.local.yaml` if it exists, and then flags disable that behaviour and just load in the specified config(s) instead?

Adding flags I think means we remove the `NODE_ENV`/`APP_ENV`-based resolution completely.

**Comment by Rugvip** [comment]
@andrewthauer Yep, definitely leaning towards `extends`, my only hesitation is that it'd be mixed in with the rest of the config at the top-level, which is unfortunate. Just realized that we've actually set a president with having `$` denote "special functions" in yaml though, both in config and the catalog. Might make sense to make it a top-level `$extends` instead.

The frontend config would be merged into a singe config if it was simple to do 😅. It's nice to get error messages that point you to the correct config, so not merging it into a single blob helps with that. We could still mostly remove duplicate values, but that's just a bit of an optimization to get page load size down a tiny bit, which we could do at some point. The big reason is that the artifact for the frontend build is a set of static files, which can then be served in any way you want, as well as the frontend being completely decoupled from the backend in the architecture. As in you're able to build plugins and serve the frontend without even deploying the backend. Because the artifact is static files, any additional config injection is always gonna be tricky. We currently have duplicated config injection logic implemented both in the backend and with `jq` in `bash` for nginx.

A possibility would be to switch to the frontend and backend being coupled by default, and use a backend config plugin or similar to expose the frontend config to the frontend. The config loader in the frontend is still pluggable, so one could revert to the existing behavior if needed. I still wanna be a bit careful though, as something that we're likely to swift towards in the future is more development towards existing backend deployments. Internally our backend(s) are a lot more heavy than what you'd want to run locally for development, so we mostly rely on the production backend endpoints, except for perhaps a few that you'd be working on. We'd have to make sure the config loading fits in alright with that model too.

Overall I'm thinking we could start out with fixing the confusion around what files are being loaded, and then perhaps see if work towards #2854 reveals some opportunities for making the loading of the frontend config a bit nicer too.

**Comment by Rugvip** [comment]
Gonna settle on multiple `--config <path>` flags, along with defaulting to the equivalent of (if they exist) `--config app-config.yaml --config app-config.local.yaml`. Which will largely behave similar to the current dev setup, but allow for explicit selection of config.

Will see if it makes sense to provide backwards compatibility for this or not. May settle for trying to identify potentially incorrect config loads and warn/error instead.
